smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/StructureGenerator.kt

```diff
@@ -30,6 +31,7 @@ class StructureGenerator(
     private val symbolProvider = ctx.symbolProvider
     private val model = ctx.model
     private val symbol = ctx.symbolProvider.toSymbol(ctx.shape)
+    private val nullableIndex = NullableIndex(ctx.model)
```

Comment:
**Suggestion**: We shouldn't need to reinstantiate the `NullabilityIndex` when it's already present in the `symbolProvider`. I suggest encapsulating the logic to determine if a member is nullable into `KotlinSymbolProvider` and then call that method from this class.

Reply:
In most places though I think we rely on the generic `SymbolProvider` rather than the actual `KotlinSymbolProvider` implementation

Reply:
I pinged Ian about this yesterday, this change will snowball into a ton of type changes across several files, I don't feel like it's a net-win to be honest, but I already have the changes so I might as well post them for you guys to decide.

Reply:
I'm fine with replacing `SymbolProvider` with `KotlinSymbolProvider` as you've posted...despite the number of changes it incurs, I think it's the right change to make.

One alternative worth discussing is retaining the previous type dependencies and just adding an extension method for `isMemberNullable` onto `SymbolProvider`:

```kotlin
fun SymbolProvider.isMemberNullable(member: MemberShape): Boolean = when (this) {
    is KotlinSymbolProvider -> isMemberNullable(member)
    else -> throw NotImplementedError("Cannot determine nullability of $member via $this")
}
```

Reply:
Synced up offline with Ian, we will keep the `NullableIndex` here for the time being.